src/lib/ptree/ptree.lua

```diff
@@ -347,25 +347,42 @@ function Manager:monitor_worker_counters(id)
    local events = inotify.recursive_directory_inventory_events(dir, cancel)
    for ev in events.get, events do
       if has_suffix(ev.name, '.counter') then
-         local stat = S.lstat(ev.name)
          local name = strip_prefix(ev.name, dir..'/')
          local qualified_name = '/'..pid..'/'..name
          local counters = self.counters[name]
-         if blacklisted(name) or not stat or stat.islnk then
+         if blacklisted(name) then
             -- Pass.
          elseif ev.kind == 'creat' then
             if not counters then
                counters = { aggregated=counter.create(name), active={},
                             rrd={}, aggregated_rrd=self:make_rrd(name),
-                            archived=ffi.new('uint64_t[1]') }
+                            archived=ffi.new('uint64_t[1]'),
+                            truename = {} }
                self.counters[name] = counters
             end
-            counters.active[pid] = counter.open(qualified_name)
-            counters.rrd[pid] = self:make_rrd(qualified_name)
+            local stat = S.lstat(ev.name)
+            if stat then
+               local truename = (stat.islnk and S.readlink(ev.name)) or ev.name
+               local is_unique = true
+               for _, aggregated in pairs(counters.truename) do
```

Comment:
Isn't `counters` the object that is specific to this path? Therefore `counters.truename` will always be `{}` here AFAIU.

Reply:
It’s a bit confusing. The `counters` map is per-counter and has submaps per pid that has the counter. This adds a submap pid->truename which (in the updated version) is checked by `sample_active_counters` to avoid summing symlinked counters multiple times.